### ISSUE TITLE
[apex] New Rule: OperationWithHighCostInLoop

### ISSUE THREAD

**Issue by tprouvot** [pmd/pmd#4675]
**Proposed Rule Name:** OperationWithHighCostInLoop

**Proposed Category:**  Performance

**Description:**
Methods Schema.getGlobalDescribe() or Schema.describeSObjects should be called once per an apex transaction and should never be called inside a loop.

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
//to avoid
Set<String> fieldNameSet = new Set<String> {'Id'};
		for (String fieldNameOrDefaultValue : fieldNameOrDefaultValueList) {
			if (Schema.getGlobalDescribe().get(objectName).getDescribe().fields.getMap().containsKey(fieldNameOrDefaultValue.trim() )) {
				fieldNameSet.add(fieldNameOrDefaultValue);
			}
		}
//correct
Set<String> fieldNameSet = new Set<String> {'Id'};
		Map<String, Schema.SObjectField> fieldMap = Schema.getGlobalDescribe().get(objectName).getDescribe().fields.getMap();
		for (String fieldNameOrDefaultValue : fieldNameOrDefaultValueList) {
			if (fieldMap.containsKey(fieldNameOrDefaultValue.trim())) {
				fieldNameSet.add(fieldNameOrDefaultValue);
			}
		}
```

**Additional Information:**
See also: [Why is Schema.describeSObjects(types) Slower Than Schema.getGlobalDescribe()?](https://salesforce.stackexchange.com/questions/218982/why-is-schema-describesobjectstypes-slower-than-schema-getglobaldescribe)

**Possible Properties:**

*   Should this rule be customizable via properties?
